### ISSUE TITLE
Default '.' adapter and Rivets Components – ES6 class methods are NOT enumerable

### ISSUE THREAD

**Issue by Leeds-eBooks** [mikeric/rivets#513]
Putting this here in lieu of a to-do list for @mikeric , @Duder-onomy and myself (and anyone else who contributes!). But if anyone at all who reads this has _any_ ideas on how to proceed, I am all ears.

For some reason, when building the class for a component in the following es6 way, the `perform` event handler does not get called:

```javascript
class MessageActions {
  constructor(data) {
    this.actions = data;
  }

  perform(event, scope) {
    console.log(this.actions);
  }
}

rivets.components['message-actions'] = {
  template: () =>
    `<button rv-each-obj="actions" rv-on-click="perform">
      TEST BUTTON
    </button>`,

  initialize: (el, data) => new MessageActions(data)
};
```

I can't work out why after several hours of debugging, all I know is what the following `console.log` returns:

```javascript
Binding.prototype.observe = function(obj, keypath, callback) {
  if (keypath === 'perform') {
    console.log(Object.keys(obj), obj, obj.perform);
    // --> ["index", "actions", "perform", ...] Object undefined
  }
  return Rivets.sightglass(obj, keypath, callback, {
    root: this.view.rootInterface,
    adapters: this.view.adapters
  });
};
```

If I define the MessageActions class like this, it works fine:

```javascript
class MessageActions {
  constructor(data) {
    this.actions = data;

    this.perform = (event, scope) => {
      console.log(this.actions);
    };
  }
}
```

and the log returns a function as its last argument instead of `undefined`.

I've looked into how Babel transpiles `class` declarations, and it uses `Object.defineProperty()`, which I guess could be the culprit, but it's not using getters or anything. Methods are defined by Babel as `enumerable: false, configurable: true, writable: true`. What's even weirder is that if you pass the object to `window` and check it manually in the console, it works:

```javascript
Binding.prototype.observe = function(obj, keypath, callback) {
  if (keypath === 'perform') {
    console.log(Object.keys(obj), obj, obj.perform); // --> returns `undefined` for the 3rd argument
    window.testobj = obj; // In the console, `window.testobj.perform` returns the function, not `undefined`
  }
  return Rivets.sightglass(obj, keypath, callback, {
    root: this.view.rootInterface,
    adapters: this.view.adapters
  });
};
```

**Comment by MiguelCastillo** [comment]
@Leeds-eBooks ES2015 methods in classes are by spec not enumerable. That was a controversial one... Some people wanted them enumerable for consistent with current behavior with standard objects, some people wanted a change...  Long story with arguments.

The point is that methods defined in classes are not enumerable.

But all that means is that Object.keys(obj) is *not* going to give back `perform` when `perform` is in the class definition vs added as a property in the constructor.  From the code you have here, I don't see where obj.perform is being called.

The code you have does not make sense.  obj.perform should be method.  Is there a chance that by the time Binding.prototype.observe got called, there is a steps that coerces the class instance to a literal object?  Perhaps somewhere in the code `merge` or something similar is being used to clone the object, which will only copy enumerables...

**Comment by Leeds-eBooks** [comment]
@MiguelCastillo Yes you're right, not enumerable, perhaps I copy/pasted it wrong. I'll edit my question now. But the strange issue still stands, and as you say, all I can think is that the class is being coerced somewhere. I've tried following the journey through the rivets/sightglass code, but it's difficult without some clear logging, and it could take hours... I'll put up a fiddle when I can.

**Comment by Leeds-eBooks** [comment]
Ok [here's a fiddle](http://codepen.io/Leeds-eBooks/pen/PqVdzK?editors=101) demonstrating the issue. The method declared on `this` in the `constructor` (called `testThis`) works fine if you click the buttons **abc** or **def**, but the method declared according to es6 (called `testMethod`) is not called if you click the buttons **123** or **456**.

**Comment by tjcrowder** [comment]
@Leeds-eBooks : Just a note, when I run that CodePen at the moment, clicking `123` or `456` works, but looking at the transpiled code that's because the version of Babel that CodePen is using is creating `MessageActions.prototype.testMethod` incorrectly, it's making it *enumerable* when it shouldn't. Current Babel doesn't do that, it makes it non-enumerable per spec. [This pen](http://codepen.io/anon/pen/JGrprp?editors=101) is your original forked and with `testMethod` added to the prototype with the correct flags (including enumerable=false, by default), and sure enough, it doesn't work anymore.

If Rivets is using `Object.keys` (which only visits own, enumerable properties), it may make sense to use [`getOwnPropertyNames`](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-object.getownpropertynames) instead, which gets own properties regardless of whether they're enumerable (provided the property name is a string, not a `Symbol`).

**Comment by Leeds-eBooks** [comment]
@tjcrowder Thanks very much for testing this out! That's a great find and goes some way to proving that this does indeed have something to do with whether or not the method is enumerable. I'm going to dig into the code now and see if I can find the problem.

The question of whether or not it needs to walk the prototype chain is an interesting and important one, I invite @Duder-onomy to weigh in here.

**Comment by Leeds-eBooks** [comment]
Ok I think I've found the culprit: https://github.com/mikeric/rivets/blob/master/src/view.coffee#L137
(and in the es6 branch: https://github.com/mikeric/rivets/blob/es6/src/view.js#L226)

I will try using `getOwnPropertyNames` instead and see if it helps.

**Comment by Leeds-eBooks** [comment]
Hmm no, not that. I'll keep digging.

**Comment by Leeds-eBooks** [comment]
It would be great if @mikeric could help out with this (i.e. finding where rivets is enumerating the component class properties), I think it would take him seconds/minutes whereas it's taking me hours!

**Comment by blikblum** [comment]
In the following pen i've got class method to be called:
http://codepen.io/blikblum/pen/yePXMz

The value of this however is the element that fired the event and not the component scope

**Comment by Leeds-eBooks** [comment]
@blikblum Wow, nice. What's the difference, what have you done differently? Read the code but can't work out what change makes this work.

**Comment by blikblum** [comment]
Seems that, as @tjcrowder  pointed, is related to the babel version codepen is using.

As i side note, the scope instance that is passed to each button click event, is a mix between the variable defined in the binding with the scope created at component `initialize`. Also the this value in the functions calls is not consistent.
See console output of http://codepen.io/blikblum/pen/obpNBQ

So IMO, component architecture in rivets needs some more work than just allowing es6 class methods

**Comment by HendrikvA** [comment]
I may found a solution to this problem. Here is a patch for rivets.js that works for me:
@@ -27,7 +27,11 @@
         return '%' + modelName + '%';
       },
       handler: function(context, ev, binding) {
-        return this.call(context, ev, binding.view.models);
+    	if(binding.model["constructor"] == null){
+    		return this.call(context, ev, binding.view.models);
+    	}else{
+    		return this.call(binding.model, ev, binding.view.models);
+    	}
       },
       configure: function(options) {
         var descriptor, key, option, value;

The idea behind this, is that, when the object, that holds the calling function, has a constructor, it's liky a es6-class. So i chaged the binding in that case...

Im not familiar with coffee so i only guess that rivets/src/rivets.coffee:54 could look like that:
  if  binding.model.constructor?
      @call binding.model, ev, binding.view.models
  else
      @call context, ev, binding.view.models
 
Of course this needs some testing, whitch i'm unable todo, because i'm very new to rivets, and don't even know wheter I'm continiue using it...

**Comment by HendrikvA** [comment]
I analized a bit further, and found out that "if(binding.model["constructor"] == null)" is always false, so the if does not make sense. This is because "context" is the HTML-Element, that the event was bound to and an HTML-Element has a constructor.
I would sugesst changeing line rivets/src/rivets.coffee:54 to:
```
@call binding.model, ev, binding.view.models
```
or 
```
@call binding.model, ev, binding.view.models, context
```
the former context would be also accessible via ev.target and i don't know if many people are using this contextchange...

**Comment by HendrikvA** [comment]
I've read the documentation, and this the handler can be changed manualy via the configuration-option: So just put:
```
rivets.configure({
	  handler: function(target, event, binding) {
	    this.call(binding.model, event, binding.view.models)
	  }
})
```
somewhere to make ES6-Callbacks possible.

I think someone should mention that in the documentation...
